tokio/docs/runtime-stats.md

```diff
@@ -0,0 +1,137 @@
+
+A small note, the term `stats` is used instead of `metrics` because we are only concerned with exposing raw data rather than methods of aggregating and emiting that data.
+
+## Motivation
+
+When developing and writing Tokio applications, there are many forms of support, be it tutorials or discord. But when running these applications in production there is not much support. Users want to understand what is happening behind the scenes. What is my runtime up too? How can I optimize my application? This RFC intends to provide a foundation to answer these questions.
+
+## Guide-level explanation
+
+Runtime stats will be exposed via a struct that is attainable via the `tokio::runtime::Handle`. Calling `Handle::stats()` will return a reference counted struct that contains raw stat values. Through this, there will be a `tokio-metrics` crate that converts these raw stats into proper aggregated metrics that can be consumed by end user metrics collection systems like the `metrics` crate.
+
+```rust=
+// Handle that can be acquired via `Runtime` or via `Handle::current`.
+let handle = Handle::current();
+let stats = handle.stats();
+
+// executor level stats.
+let executor = stats.executor();
+
+// per-worker stats via the executor.
+for worker in executor.workers() {
+    println!("{}", worker.steal_count());
+}
+
+println!("{}", executor.global_queue_depth());
+
+// io driver stats.
+let io = stats.io();
+
+println!("{}", io.sources_count());
+
+let timer = stats.timer();
+
+println!("{}", timer.entry_count());
+```
+
+### Executor
+
+The handle will provide stats for each worker thread or in the case of the single threaded executor will provide a single worker. This provides a detailed view into what each worker is doing. Providing the ability for the `tokio-metrics` crate to expose the stats aggregated as a single metric or as a per-worker metric.
+
+The values will be updated in batch from the executor to avoid needing to stream the data on every action. This should amortize the cost by only needing to emit stats at a specific executor wall clock time. Where the executor wall clock time is determinetd by a single executor tick rather than actual system time. This allows the collectors to observe the time and the stats to determine how long certain executor cycles took. This removes the need to acquire the time during executor cycles.
+
+Each worker will expose these stats, updated in batches:
+
+- Amount of futures executed
+- Amount of work steal attempts
+- Amount of parks
+- Amount of times local queue overflowed into the global queue
+- Max local queue depth
+- Min local queue depth
+- Avg local queue depth
+- Queue depth at time of batch emission
+- Amount of executor ticks (loop iterations)
```

Comment:
If my understanding is correct, you intend to provide information about how long executors spend busy in tasks by allowing collectors to observe the number of ticks which occur in a known time interval. However, I think you also need to measure the amount of time spent parked in order to avoid counting parked time as "busy" time.